Add project-specific script repositories to Sindarin and make 'Create Command' button working

FlavienVolant opened the pull request:
Add project-specific script repositories to Sindarin

https://github.com/user-attachments/assets/0e3e2b05-65e5-4f13-aa24-a66099a35ea0


 'Create Command' button

https://github.com/user-attachments/assets/ead21766-93e6-4067-9c30-76e38f357699

FlavienVolant committed:
Refactor script repository system to support project-specific repositories

- Made `StSindarinDebuggerScriptRepository` an abstract class with `repositoryName` to be implemented by subclasses
- Introduced `SindarinDefaultScriptRepository` to manage Pharo's default scripts
- Updated `StSindarinDebuggerScriptRepositoryPresenter`:
  - Added a dropdown listing all subclasses of `StSindarinDebuggerScriptRepository` to select the active repository
  - Added `loadScript` and `saveScript` to work with the selected repository
- Updated `StSindarinDebuggerScriptingPresenter` to support changes from `StSindarinDebuggerScriptRepositoryPresenter`

FlavienVolant committed:
Refactor repository selection into dedicated chooser presenter
- Added StSindarinDebuggerScriptRepositoryChooserPresenter as a subclass of SpDropListPresenter to handle repository selection
- Refactored StSindarinDebuggerScriptRepositoryPresenter to receive the repository as an attribute (removed internal drop list)
- Integrated the new chooser into StSindarinDebuggerScriptingPresenter to manage repository switching

FlavienVolant committed:
Improves clarity by distinguishing between repository selection and script selection
- Renamed StSindarinDebuggerScriptRepositoryChooserPresenter to repositoryChooser
- Renamed StSindarinDebuggerScriptRepositoryPresenter to scriptChooser

FlavienVolant committed:
move acces to save / load script from the scriptChooser to the scriptRepositoryChooser

FlavienVolant committed:
rename initializeScriptRepository to initializeScriptChooser

FlavienVolant committed:
Add StSindarinDebuggerCreateCommandPresenter for command creation UI

- Created StSindarinDebuggerCreateCommandPresenter to let users choose a description and an icon for a new command (command creation logic not implemented yet).
- Updated StSindarinDebuggerScriptingPresenter to use the new presenter when the user clicks on "create command".

FlavienVolant committed:
- createCommand create the class without any methods (todo)
- SindarinDefaultScriptRepository is moved to NewTools-Sindarin-Script wich required to add thi package to the baseline

FlavienVolant committed:
temporary remove sindarin :=

FlavienVolant committed:
- SindarinCommand remove Override of context: 
- StSindarinDebuggerCreateCommand createCommand done

FlavienVolant committed:
- createCommand method calls createMenu to automatically create the debugger menu

FlavienVolant committed:
- fix command name
- fix load layout

FlavienVolant commented:
@StevenCostiou

jecisc commented on src/NewTools-Sindarin-Tools/StSindarinDebuggerCreateCommandPresenter.class.st:
```diff
@@ -0,0 +1,153 @@
+
+{ #category : 'actions' }
+StSindarinDebuggerCreateCommandPresenter >> createMenu [
+	StDebugger class compile: ('buildSindarin{2}ExtentionCommandsGroupWith: stDebuggerInstance forRoot: rootCommandGroup
```
Is it normal to have this?

If yes, it might be nice to add a comment to explain why we have a method compiling menus :)

Ducasse replied:
I do not think that relying on the compile is good. Especially since we will probably change this in Toplo

FlavienVolant replied:
The goal is to generate a command group for the debugger.  To do that I need to add a class side method to StDebugger. I followed 5.4 "Adding a new menu extending the debugger" from https://github.com/The-Pharo-Debugger-Book-Chapters/Sindarin-Scriptable-Debugger.

If there is a better way to add a method without using compile, I'm open to using it.

StevenCostiou replied:
We need to compile code, so at some point we have to call "compile:". 
I do not understand the relation between compiling code and Toplo...?

I would say though the "createMenu" is not a correct method name.

jecisc commented on src/NewTools-Sindarin-Tools/StSindarinDebuggerCreateCommandPresenter.class.st:
```diff
@@ -0,0 +1,153 @@
+	
+	(self debuggerToolbarCommandGroupFor: rootCommandGroup) register: toolbarCommandGroup.
+	
+	commands do: [ :c | toolbarCommandGroup register: c ].' format: {projectName. projectName asCamelCase . package})
```
```suggestion
	commands do: [ :command | toolbarCommandGroup register: command ].' format: {projectName. projectName asCamelCase . package})
```

In Smalltalk we try to avoid abbreviation. Since we have completion tool it's better to have a readable code than to gain 2sec to type the code

jecisc commented on src/NewTools-Sindarin-Tools/StSindarinDebuggerScriptingPresenter.class.st:
```diff
@@ -35,7 +37,16 @@ StSindarinDebuggerScriptingPresenter >> accept: aVisitor [
 { #category : 'actions' }
 StSindarinDebuggerScriptingPresenter >> createCommandFromScript [
-	self flag: 'todo'
+	self flag: 'todo'.
```
I see the code was there already, but is the todo still needed? If yes, it'd be nice to add a comment to say what needs to be done

FlavienVolant replied:
I created this pr for @StevenCostiou to review my code but it's not finished yet. I left the todo because there is a display issue when I generate a new command group.

jecisc reviewed (commented):
I did a few comments

Also do not hesitate to format methods using cmd/ctrl + shift + f.

It removes for example useless spaces or dots

FlavienVolant committed:
Update src/NewTools-Sindarin-Tools/StSindarinDebuggerCreateCommandPresenter.class.st

Co-authored-by: CyrilFerlicot <cyril@ferlicot.fr>

FlavienVolant committed:
- Added `deleteCommand` and `deleteMenu` methods to `StSindarinDebuggerCreateCommandPresenter` to allow removal of user-created commands and groups.
- Removed the TODO flag from `createCommandFromScript`.
- Implemented `removeScriptCommand`

FlavienVolant committed:
Include project name in generated command class names

FlavienVolant committed:
format methods using cmd/ctrl + shift + f

FlavienVolant committed:
Add documentation to script repository and dynamically generated commands

FlavienVolant committed:
- expand  toolbar layout
- rename StSindarinDebuggerCreateCommandPresenter to StSindarinDebuggerCommandPresenter
- rename whenCancelButtonDo to whenCancelDo
- rename whenLoadButtonDo to whenLoadDo
- rename createCommand and createMenu to generateCommand and generateMenu
